### PR TITLE
JENKINS-65577 Add admin monitor recommending java 11 for controller

### PR DESCRIPTION
**Pull request by timja**
See [JENKINS-65577](https://issues.jenkins-ci.org/browse/JENKINS-65577).
See https://github.com/jenkins-infra/jenkins.io/pull/4335

image

### Proposed changelog entries

* Entry 1: JENKINS-65577, Recommend running on Java 11
* ...

### Proposed upgrade guidelines

N/A

### Submitter checklist

- [ ] (If applicable) Jira issue is well described
- [ ] Changelog entries and upgrade guidelines are appropriate for the audience affected by the change (users or developer, depending on the change). [Examples](https://github.com/jenkins-infra/jenkins.io/blob/master/content/_data/changelogs/weekly.yml)
  * Fill-in the `Proposed changelog entries` section only if there are breaking changes or other changes which may require extra steps from users during the upgrade
- [ ] Appropriate autotests or explanation to why this change has no tests
- [ ] For dependency updates: links to external changelogs and, if possible, full diffs

### Desired reviewers

@mention

### Maintainer checklist

Before the changes are marked as `ready-for-merge`: 

- [ ] There are at least 2 approvals for the pull request and no outstanding requests for change
- [ ] Conversations in the pull request are over OR it is explicit that a reviewer does not block the change
- [ ] Changelog entries in the PR title and/or `Proposed changelog entries` are correct
- [ ] Proper changelog labels are set so that the changelog can be generated automatically
- [ ] If the change needs additional upgrade steps from users, `upgrade-guide-needed` label is set and there is a `Proposed upgrade guidelines` section in the PR title. ([example](https://github.com/jenkinsci/jenkins/pull/4387))
- [ ] If it would make sense to backport the change to LTS, a Jira issue must exist, be a _Bug_ or _Improvement_, and be labeled as `lts-candidate` to be considered (see [query](https://issues.jenkins-ci.org/issues/?filter=12146)).

### CODE REVIEW

File: core/src/main/resources/jenkins/monitor/JavaLevelAdminMonitor/message.properties
Comment (timja) [inline core/src/main/resources/jenkins/monitor/JavaLevelAdminMonitor/message.properties:1]: please suggest any wording improvements

**Comment by timja** [comment]
Requested all people who participated in the mailing list thread

File: core/src/main/resources/jenkins/monitor/JavaLevelAdminMonitor/message.properties
Comment (res0nance) [inline core/src/main/resources/jenkins/monitor/JavaLevelAdminMonitor/message.properties:2]: Are we sure we will be dropping it in 2021? Perhaps when we have an appropriate timeline this should be updated?

File: core/src/main/resources/jenkins/monitor/JavaLevelAdminMonitor/message.properties
Comment (jvz) [reply core/src/main/resources/jenkins/monitor/JavaLevelAdminMonitor/message.properties:1]: Maybe something like Jenkins runs better on Java 11, too? Some sort of treat.

**Review by slide — APPROVED** [review]

**Comment by daniel-beck** [comment]
On-holding until all Docker images and packages use Java 11 by default (otherwise this would be kind of silly).

**Comment by timja** [comment]
> On-holding until all Docker images and packages use Java 11 by default, as well as all user documentation on jenkins.io is updated (otherwise this would be kind of silly).

User documentation was updated months ago, may not be 100%, but the installation guides were done.

Packages don't depend on java to my knowledge because of distribution packaging issues, the user has to manage java, and that's handled in installation guide.

Docker images default can happen when blog posts / whatever is prepared.

We can drop the 'java 8 jenkins dropping support' line for now till we have more clarity on when if that helps.

Shipping this before the docker 'breaking change' of image defaults can nudge users to upgrade before that happens. The more people who upgrade before any 'breaking change' of images or Jenkins actually dropping support the better I would have thought.

TLDR: Apart from polishing the wording here and possibly in some documentation I don't see a reason, I'm not saying merge this tomorrow, but I see real value in getting this out to users before any 'breaking changes' happen.

**Comment by daniel-beck** [comment]
> User documentation was updated months ago, may not be 100%, but the installation guides were done.

I added this to my comment after seeing Java 8 referenced on https://www.jenkins.io/doc/book/installing/linux/#long-term-support-release-3

Some documentation like https://www.jenkins.io/doc/administration/requirements/java/#running-jenkins still uses phrasing suggesting that Java 11 not the recommended/default version. Nobody cares about 2.164 anymore.

> Docker images default can happen when blog posts / whatever is prepared.

My point is that users should not see this message if they install the Jenkins release this gets merged into, following the then-current instructions. AFAIUI, they would with all the Docker images, probably some of the other packages as well.

**Comment by timja** [comment]
> I wonder whether we can identify that we're running inside Docker, and show specific actionable advice to those users. The paragraph touching on Docker in jenkins.io/doc/administration/requirements/upgrade-java-guidelines is not discoverable enough IMO.

should be possible

> I added this to my comment after seeing Java 8 referenced on jenkins.io/doc/book/installing/linux/#long-term-support-release-3

Created https://github.com/jenkins-infra/jenkins.io/issues/4336

**Comment by daniel-beck** [comment]
> Again documentation was updated months ago at least on the website, any thing missed is a bug and issues / fixes are greatly appreciated

Took me less than two minutes to find occurrences on https://www.jenkins.io/doc/book/installing/other/ (both OpenIndiana and Solaris sections) using `[^0-9]8[^0-9]` on `/doc/book/`, which to me seems like a very basic strategy to discover wrong version recommendations 🤷

**Review by MarkEWaite — COMMENTED** [review]
Possible phrasing change?

File: core/src/main/resources/jenkins/monitor/JavaLevelAdminMonitor/message.properties
Comment (timja) [reply core/src/main/resources/jenkins/monitor/JavaLevelAdminMonitor/message.properties]: I don't think we use the jenkins project in jenkins anywhere else, doesn't make sense to me

File: core/src/main/java/jenkins/monitor/JavaLevelAdminMonitor.java
Comment (daniel-beck) [inline core/src/main/java/jenkins/monitor/JavaLevelAdminMonitor.java:60]: Unnecessary, admin monitors get that via `StaplerProxy#getTarget`.

File: core/src/main/java/jenkins/monitor/JavaLevelAdminMonitor.java
Comment (daniel-beck) [inline core/src/main/java/jenkins/monitor/JavaLevelAdminMonitor.java:36]: Should be `java.specification.version` I think, but no need to reinvent the wheel here, `jenkins.util.java.JavaUtils` probably has what you need.

File: core/src/main/resources/jenkins/monitor/JavaLevelAdminMonitor/message.jelly
Comment (daniel-beck) [inline core/src/main/resources/jenkins/monitor/JavaLevelAdminMonitor/message.jelly:4]: We don't have any admin monitors with headers in core. Does it make sense to add this to the presentation zoo?

File: core/src/main/java/jenkins/monitor/JavaLevelAdminMonitor.java
Comment (timja) [reply core/src/main/java/jenkins/monitor/JavaLevelAdminMonitor.java:60]: sure, this definitely came from a different admin monitor 😄

File: core/src/main/java/jenkins/monitor/JavaLevelAdminMonitor.java
Comment (daniel-beck) [inline core/src/main/java/jenkins/monitor/JavaLevelAdminMonitor.java:20]: `@Symbol`?

File: core/src/main/java/jenkins/monitor/JavaLevelAdminMonitor.java
Comment (timja) [reply core/src/main/java/jenkins/monitor/JavaLevelAdminMonitor.java]: no longer redundant as I enabled system read on it (was originally intended)

File: core/src/main/java/jenkins/monitor/JavaVersionRecommendationAdminMonitor.java
Comment (daniel-beck) [inline core/src/main/java/jenkins/monitor/JavaVersionRecommendationAdminMonitor.java:1]: Missing license header

File: core/src/main/resources/jenkins/monitor/JavaVersionRecommendationAdminMonitor/message.jelly
Comment (daniel-beck) [inline core/src/main/resources/jenkins/monitor/JavaVersionRecommendationAdminMonitor/message.jelly:7]: Nicely split up 👍

**Review by daniel-beck — APPROVED** [review]
Looks good from an implementation POV.

(Still not convinced this is the right approach to suggest upgrades.)

**Review by oleg-nenashev — APPROVED** [review]
I would recommend a killswitch SystemProperty which allows disabling this monitor. Speaking for a friend vendor company that still does not support Java 11. Not blocking

**Comment by timja** [comment]
> I would recommend a killswitch SystemProperty which allows disabling this monitor. Speaking for a friend vendor company that still does not support Java 11. Not blocking

Is that so a vendor can exclude it rather than just automatically adding it to the disabled monitor list?

**Comment by daniel-beck** [comment]
We patch Jenkins anyway; I'm sure we'll figure out how to comment out https://github.com/jenkinsci/jenkins/pull/5474/files#diff-221cda3ae1e01a23844e3c1dcb7fbb9c98e6bb928d5426dbff5b4b95c7367a2cR44

**Review by res0nance — APPROVED** [review]

**Comment by timja** [comment]
This PR is now ready for merge, after ~24 hours, we will merge it if there's no negative feedback.

Thanks!

**Pull request merged**
